choreographer/src/main/java/eu/arrowhead/core/choreographer/executor/ExecutorSelector.java

```diff
@@ -0,0 +1,172 @@
+			return null;
+		}
+		
+		final Map<Long,ChoreographerExecutorServiceInfoResponseDTO> executorServiceInfos = collectExecutorServiceInfos(potentials, serviceDefinition, minVersion_, maxVersion_);
+		potentials = prioritizationStrategy.priorize(potentials, executorServiceInfos);
+		
+		return selectVerifyAndInitFirstAvailable(potentials, executorServiceInfos, init);
+	}
+	
+	//=================================================================================================
+	// assistant methods
+	
+	//-------------------------------------------------------------------------------------------------
+	private List<ChoreographerExecutor> filterOutLockedAndExcludedExecutors(final List<ChoreographerExecutor> original, final List<ChoreographerExecutor> exclusions) {
+		logger.debug("filterOutLockedAndExcludedExecutors started...");
+		
+		final Set<Long> excudedIds = exclusions.stream().map(e -> e.getId()).collect(Collectors.toSet());
+		
+		final List<ChoreographerExecutor> filtered = new ArrayList<>(original.size());
+		for (final ChoreographerExecutor executor : original) {
+			if (!executor.isLocked() && !excudedIds.contains(executor.getId())) {
+				filtered.add(executor);
+			}
+		}			
+		return filtered;		
+	}
+	
+	//-------------------------------------------------------------------------------------------------
+	private Map<Long,ChoreographerExecutorServiceInfoResponseDTO> collectExecutorServiceInfos(final List<ChoreographerExecutor> potentials, final String serviceDefinition,
+																							  				   final int minVersion, final int maxVersion) {
+		logger.debug("collectExecutorServiceInfos started...");
+		
+		final Map<Long,ChoreographerExecutorServiceInfoResponseDTO> collected = new HashMap<>(potentials.size());
+		for (final ChoreographerExecutor executor : potentials) {
+			try {
+				final ChoreographerExecutorServiceInfoResponseDTO info = driver.queryExecutorServiceInfo(executor.getAddress(), executor.getPort(), executor.getBaseUri(),
+																										 serviceDefinition, minVersion, maxVersion);
+				collected.put(executor.getId(), info);
+				
+			} catch (final Exception ex) {
+				logger.debug(ex.getMessage());
+			}
+		}
+		return collected;
+	}
+	
+	//-------------------------------------------------------------------------------------------------
+	private ChoreographerExecutor selectVerifyAndInitFirstAvailable(final List<ChoreographerExecutor> potentials, final Map<Long,ChoreographerExecutorServiceInfoResponseDTO> executorServiceInfos,
+																	final boolean init) {
+		logger.debug("selectVerifyAndInitFirstAvailable started...");
+		
+		if (potentials.isEmpty()) {
+			return null;
+		}
+		
+		for (final ChoreographerExecutor potential : potentials) {
+			final Optional<ChoreographerExecutor> optional = executorDBService.getExecutorOptionalById(potential.getId()); //refreshing from DB
+			if (optional.isEmpty()) {
+				continue;
+			}
+			final ChoreographerExecutor executor = optional.get();
```

Comment:
you should also continue if the executor's id is not in the map (means something is happened during the information gathering, i.e. executor was unreachable)